### DISCUSSION TITLE
Elasticsearch: add support for calendar_interval

### DISCUSSION THREAD

**Discussion by gabor** [grafana/grafana#55034] [Observability]
in the past, with using Elasticsearch7.x, it was possible to trigger elastic's `calendar_interval` handling. this support was accidental (see https://github.com/grafana/grafana/issues/19911#issuecomment-971652104), but it's currently not possible anymore, since grafana9.1.0 .

we should identify the use-cases where `calendar_interval` is needed, and add support for it. 

it's not clear yet which values are needed. technically speaking, elasticsearch only allows the following fixed values to be sent for this setting:
- 1 minute
- 1 hour
- 1 day
- 1 week
- 1 month
- 1 quarter
- 1 year

i'm not sure if we need supporting all of them, and how exactly they are used by grafana users. we should collect feedback about their use cases.

**Comment by sigriffin** [comment]
I have a dashboard with week by week statistics in a table used for trend analysis etc. This has been broken by the recent change.

I cannot use 7d as this then shows the previous 7 days rather than a calendar week.

**Reply by gabor** [reply]
@sigriffin there's a workaround to align the start and end timestamps on weekend boundaries, you could do something like this:
- plan is to see the last 28days, calendar-week aligned
- set the timerange to: `from=now-28d/w`, `to=now/w` (https://grafana.com/docs/grafana/v9.0/dashboards/time-range-controls/#time-units-and-relative-ranges)
- set the interval to `7d`.

does this work? (this is not the nicest solution of course, but it may help you)

**Reply by sigriffin** [reply]
![weekly rows](https://user-images.githubusercontent.com/30828869/191502712-fb0db56e-7d55-431c-9d8c-c1abc7dcf5a7.png)
![query](https://user-images.githubusercontent.com/30828869/191503981-475b2a3e-4c9a-4920-b9bc-5934be87d0b9.png)
![timerange](https://user-images.githubusercontent.com/30828869/191504190-e168b759-c88b-44fd-82ac-084bf94753e1.png)
I tried using the timerange and interval as advised, however this is the result when just using a min(timestamp) for the metric as a simple example. As you can see the dates are aligning to week ending "today" rather than a weekend boundary. The actual data in the first row does start on a Sunday even if the timestamp doesn't but all subsequent rows then do not.

**Comment by akosiaris** [comment]
Hi,

Thanks for starting this discussion. Adding my use case as asked in https://github.com/grafana/grafana/pull/50297#issuecomment-1243795512

What I have is hourly data (Electricity prices) for which I am calculating a calendar month `avg()`. With `calendar_interval`, this works fine (I can cross check it with other published sources and results match). I can't see how I can work around not having that (I downgraded back to `9.0.x` because of it). Handling the 30 or 31 days of individual months, not to mention the erratic February's 28 or 29 days is something I 'd much rather offload to Elasticsearch. 

I also have some week uses cases too (weekly covid data) but for a variety of reasons (most notably I don't care that much about data accuracy for my use cases), switching to 7d is an adequate workaround.

Forgive my naivete, but since this was working before 9.1.0, and assuming there wasn't any date time breaking changes in internal Grafana's APIs, if one switched the interval box to a dropdown allowing to choose between `fixed_interval` and `calendar_interval` would it work? 

![image](https://user-images.githubusercontent.com/814130/194046044-4c3f2197-ad4f-4ef9-adb1-995420f69e87.png)

**Reply by jlaudenbach** [reply]
I don't think a dropdown is necessary.
Since each value implicitly contains whether it is a `fixed_interval` or `calendar_interval`.
So a lookup table in the code would be enough to adjust the elasticseach request.

**Reply by akosiaris** [reply]
Elasticsearch went down that road (implicit) at the beginning with the interval field. And in 7.2 they deprecated that. Just pasting  from https://www.elastic.co/guide/en/elasticsearch/reference/7.17/search-aggregations-bucket-datehistogram-aggregation.html#calendar_and_fixed_intervals

```
 [7.2] Deprecated in 7.2. interval field is deprecated Historically both calendar and fixed intervals were configured in a single interval field, which led to confusing semantics. Specifying 1d would be assumed as a calendar-aware time, whereas 2d would be interpreted as fixed time. To get "one day" of fixed time, the user would need to specify the next smaller unit (in this case, 24h).

This combined behavior was often unknown to users, and even when knowledgeable about the behavior it was difficult to use and confusing.

This behavior has been deprecated in favor of two new, explicit fields: calendar_interval and fixed_interval.

By forcing a choice between calendar and intervals up front, the semantics of the interval are clear to the user immediately and there is no ambiguity. The old interval field will be removed in the future.
```

Note how `2d` vs `1d` had a different meaning, confusing people. Also note that even 1m is NOT always 60s (blame leap seconds), nor is 1h always  3600 seconds. Grafana would only end up re-implementing that confusion if it went with an implicit choice. As Elasticsearch found out the hard way, it's better to force the user to choose. 

The question I 'd pose instead is more like "Should `calendar_interval` or `fixed_interval` be the default one offered"? There are 3 units (minutes, hours, days) for which both settings are valid, which one is preferable? 

And should it warn the user if they override and choose something that is unsupported by Elasticsearch (e.g. `1 month` while having `fixed_interval` selected)? or just raise the error from Elasticsearch to the user? Or just switching to the proper setting (`calendar_interval` in this case, disallowing overrides?).

My take would be:

* Seconds -> always fixed_interval (calendar interval doesn't support seconds anyway)
* minute, hour, day -> default to fixed interval, but allow users to switch to calendar_interval
* week, month, quarter, year -> always calendar_interval

Just surface Elasticsearch's error when a user submit something that's unsupported, e.g. 2d in calendar_interval

**Reply by jlaudenbach** [reply]
Thank you @akosiaris for this very good explanation!
Now I fully share our opinion that best would be to implement a drop down for choice between interval type.

just one addition, I think this drop down than also has to be added to elastic search data source field `Min time interval`

**Reply by akosiaris** [reply]
Hmmm, hadn't thought of the `Min time interval` thing. It does sounds reasonable that a similar logic if followed there

**Reply by jlaudenbach** [reply]
Another thought:
Would a simple configuration instead of a drop down be possible as a first implementation?
In the configuration you then specify `calendar_interval` or `fixed_interval` and this applies to all fields. 

Do you see any problem in that solution?
Maybe `1d` VS `2d` could be a problem again, since 2d could not be a  `calendar_interval` , right?

**Comment by akosiaris** [comment]
For what is worth, I can also see a use case for the `1d` calendar interval. Never had to check it up to now, but if you live in a country that honors DST, twice per year the day isn't 24 hours long. Elasticsearch does account for that per docs

```
 day, 1d
    All days begin at the earliest possible time, which is usually 00:00:00 (midnight). One day (1d) is the interval between the start of the day and the start of the following day in the specified time zone, compensating for any intervening time changes. 
```

**Comment by jlaudenbach** [comment]
I would also welcome the introduction of a new option for `fixed_interval` and `calendar_interval`.
Until now, we have often made evaluations on a monthly level in Grafana (Elasticsearch). I think this case is not unusual. Here the workaround with 30/31 days is not a solution. 
After the update to version 9.2 this is no longer possible.

As use case think just about any metric you want to see the changes for each month (e.g. group in bar chart or in a table)

**Comment by airforceguy** [comment]
I work in healthcare domain, and I have a requirement to show the trend of appointments booked in **calendar month** (not 30d/31d) . In 9.0.0 it was working and now as I have upgraded it is breaking.  I vote in favour of providing option for user to choose fixed_interval and calendar_interval. 
Is there any quick Workaround to use calendar_interval as default instead of fixed_interval ?

**Reply by jlaudenbach** [reply]
Hi @airforceguy,
The only way I have found is to add the date as a new field year-month (as a keyword) to the index.
Then you can make a table or bar chart and group by this field.
It works but only if you have control over your data source and you will lose all the benefits of having a date field.

**Reply by airforceguy** [reply]
Hi @jlaudenbach,
Thanks for the reply. I don't have control on the data source

**Comment by houdang** [comment]
Are there any news on this topic? We changed our system today and this is a deal breaker and might stop us from working with Grafana. We need to calculate and group data by month, and currently it is messing up the days. This ways the business can't use the dashboards anymore.

As suggested we need an option to also set "calendar_interval". As jlaudenbach mentioned the case is not unusal and happens very often in doing reports. Does the enterprise version maybe support this ?

**Comment by matyax** [comment]
https://github.com/grafana/grafana/pull/75459 has been merged, so it will be possible to use calendar intervals (1w, 1M, 1q, 1y) in the interval field:

![imagen](https://github.com/grafana/grafana/assets/1069378/c0d13863-75a5-4d20-8b36-8a2734a66b11)

Available in Grafana 10.2

**Comment by matyax** [comment]
Support was added. We will follow up this with some enhancements to smooth some rough edges, but considering this done.

**Comment by akosiaris** [comment]
@matyax many thanks for this! I ll test out soon enough using the `main` tag of docker images on docker hub.

**Comment by akosiaris** [comment]
Just tested, works quite satisfactory, thanks for your work on this!

**Reply by matyax** [reply]
Amazing! Thank you!
